src/main/java/com/google/collegeplanner/servlets/DatastoreServlet.java

```diff
@@ -0,0 +1,179 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.http.client.utils.URIBuilder;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+
+/** Queries the UMD API and downloads the data to datastore. */
+@WebServlet("/api/download")
+public class DatastoreServlet extends BaseServlet {
+  final int PAGE_LIMIT = 200;
+
+  DatastoreService datastore;
+
+  public DatastoreServlet() {
+    this(DatastoreServiceFactory.getDatastoreService(), new ApiUtil());
+  }
+
+  public DatastoreServlet(DatastoreService datastore, ApiUtil apiUtil) {
+    super(apiUtil);
+    this.datastore = datastore;
+  }
+
+  /*
+   * Downloads all course data to datastore.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Loop through the pages.
+    // We don't know how many pages there are beforehand. We stop looping when the new page
+    // doesn't have any results.
+    int page = 1;
+    do {
+      URI uri;
+      try {
+        URIBuilder builder = new URIBuilder("https://api.umd.io/v1/courses");
+        builder.setParameter("page", Integer.toString(page));
+        uri = builder.build();
+      } catch (URISyntaxException e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+      JSONArray coursesArray = apiUtil.getJsonArray(uri);
+      if (coursesArray == null) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      } else if (coursesArray.size() == 0) {
+        // Successful. The default status code response is 200.
+        return;
+      }
+      addCourses(coursesArray, response);
+    } while (page++ < PAGE_LIMIT);
+  }
+
+  /**
+   * Creates a Course object from a JSONArray.
+   * @param coursesArray The course json from the UMD API.
+   * @param response The HttpServletResponse object.
+   */
+  private void addCourses(JSONArray coursesArray, HttpServletResponse response) throws IOException {
+    if (coursesArray == null) {
+      return;
+    }
+
+    // Loop through each page's courses.
+    for (Object jsonObject : coursesArray) {
+      JSONObject courseJson = (JSONObject) jsonObject;
+
+      // Create a Course object from the JSONObject for parsing.
+      Course course;
+      try {
+        course = new Course(courseJson);
+      } catch (Exception e) {
+        respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+        return;
+      }
+
+      // Check if the entity already exists in datastore, and modify it if it does.
+      Query query = new Query("Course").setFilter(
```

Comment:
just double checking - did you create an index on course_id (and/or is it a primary/secondary key)? otherwise, this could be a linear search over all your courses.

Reply:
no I don't have an index yet. i'll need to look into how to make one

Reply:
https://cloud.google.com/datastore/docs/concepts/indexes

"By default, a Datastore mode database automatically predefines an index for each property of each entity kind. These single property indexes are suitable for simple types of queries."

Aren't all entity properties indexed by default though?

Reply:
Ah ok, good to know. Should be good then.